streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java

```diff
@@ -0,0 +1,471 @@
+                .map(KafkaStreamsState::processId)
+                .collect(Collectors.toSet());
+            final ProcessId bestClient = assignmentState.findBestClientForTask(taskId, candidateClients);
+            assignmentState.finalizeAssignment(taskId, bestClient, AssignedTask.Type.ACTIVE);
+        }
+    }
+
+    private static void assignStandby(final ApplicationState applicationState,
+                                      final AssignmentState assignmentState) {
+        final Set<TaskInfo> statefulTasks = applicationState.allTasks().stream()
+            .filter(TaskInfo::isStateful)
+            .collect(Collectors.toSet());
+        final int numStandbyReplicas = applicationState.assignmentConfigs().numStandbyReplicas();
+        for (final TaskInfo task : statefulTasks) {
+            for (int i = 0; i < numStandbyReplicas; i++) {
+                final Set<ProcessId> candidateClients = assignmentState.findClientsWithoutAssignedTask(task.id());
+                if (candidateClients.isEmpty()) {
+                    LOG.warn("Unable to assign {} of {} standby tasks for task [{}]. " +
+                             "There is not enough available capacity. You should " +
+                             "increase the number of threads and/or application instances " +
+                             "to maintain the requested number of standby replicas.",
+                        numStandbyReplicas - i,
+                        numStandbyReplicas, task.id());
+                    break;
+                }
+
+                final ProcessId bestClient = assignmentState.findBestClientForTask(task.id(), candidateClients);
+                assignmentState.finalizeAssignment(task.id(), bestClient, AssignedTask.Type.STANDBY);
+            }
+        }
+    }
+
+    private static Map<TaskId, ProcessId> mapPreviousActiveTasks(final Map<ProcessId, KafkaStreamsState> clients) {
+        final Map<TaskId, ProcessId> previousActiveTasks = new HashMap<>();
+        for (final KafkaStreamsState client : clients.values()) {
+            for (final TaskId taskId : client.previousActiveTasks()) {
+                previousActiveTasks.put(taskId, client.processId());
+            }
+        }
+        return previousActiveTasks;
+    }
+
+    private static Map<TaskId, Set<ProcessId>> mapPreviousStandbyTasks(final Map<ProcessId, KafkaStreamsState> clients) {
+        final Map<TaskId, Set<ProcessId>> previousStandbyTasks = new HashMap<>();
+        for (final KafkaStreamsState client : clients.values()) {
+            for (final TaskId taskId : client.previousActiveTasks()) {
+                previousStandbyTasks.computeIfAbsent(taskId, k -> new HashSet<>());
+                previousStandbyTasks.get(taskId).add(client.processId());
+            }
+        }
+        return previousStandbyTasks;
+    }
+
+    private static int computeStreamThreadCount(final Collection<KafkaStreamsState> clients) {
```

Comment:
nit: rename to `computeTotalProcessingThreads`

(Technically StreamThread is correct but we want to start using "processing thread" vs "consumer" to be precise when we mean one vs the other, since we have upcoming plans to decouple these. Whereas right now "StreamThread" can mean either. In case you're wondering why I'm being pedantic)